hsds/dset_lib.py

```diff
@@ -0,0 +1,272 @@
+
+    root_key = getS3Key(dset_id)
+    log.debug(f"got root_key: {root_key}")
+
+    if not root_key.endswith("/.dataset.json"):
+        raise ValueError("unexpected root key")
+
+    root_prefix = root_key[: -(len(".dataset.json"))]
+
+    log.debug(f"scanRoot - using prefix: {root_prefix}")
+
+    kwargs = {
+        "prefix": root_prefix,
+        "include_stats": False,
+        "bucket": bucket,
+    }
+    s3keys = await getStorKeys(app, **kwargs)
+
+    # getStoreKeys will pick up the dataset.json as well,
+    # so go through and discard
+    chunk_ids = []
+    for s3key in s3keys:
+        if s3key.endswith("json"):
+            # ignore metadata items
+            continue
+        try:
+            chunk_id = getObjId(s3key)
+        except ValueError:
+            log.warn(f"ignoring s3key: {s3key}")
+            continue
+        chunk_ids.append(chunk_id)
+
+    log.debug(f"getAllocattedChunkIds - got {len(chunk_ids)} ids")
+    return chunk_ids
+
+
+async def reduceShape(app, dset_json, shape_update, bucket=None):
+    """ Given an existing dataset and a new shape,
+        Reinitialize and edge chunks and delete any chunks
+        that fall entirely out of the new shape region """
+
+    dset_id = dset_json["id"]
+    log.info(f"reduceShape for {dset_id} to {shape_update}")
+
+    # get the current shape dims
+    shape_orig = dset_json["shape"]
+    if shape_orig["class"] != "H5S_SIMPLE":
+        raise ValueError("reduceShape can only be called on simple datasets")
+    dims = shape_orig["dims"]
+    rank = len(dims)
+
+    # get the fill value
+    arr = getFillValue(dset_json)
+
+    # and the chunk layout
+    layout = tuple(getChunkLayout(dset_json))
+    log.debug(f"got layout: {layout}")
+
+    # get all chunk ids for chunks that have been allocated
+    chunk_ids = await getAllocatedChunkIds(app, dset_id, bucket=bucket)
+    chunk_ids.sort()
+
+    log.debug(f"got chunkIds: {chunk_ids}")
+
+    # separate ids into three groups:
+    #   A: those are entirely inside the new shape region - no action needed
+    #   B: those that overlap the new shape - will need the edge portion reinitialized
+    #   C: those that are entirely outside the new shape - will need to be deleted
+
+    delete_ids = []  # chunk ids for chunk that that will need to be deleted
+    update_ids = []  # chunk ids for chunks that will need to be reinitialized
+
+    for chunk_id in chunk_ids:
+        log.debug(f"chunk_id: {chunk_id}")
+        chunk_coord = getChunkCoordinate(chunk_id, layout)
+        log.debug(f"chunk_coord: {chunk_coord}")
+        skip = True
+        for i in range(rank):
+            if chunk_coord[i] + layout[i] > shape_update[i]:
+                skip = False
+                break
+        if skip:
+            log.debug(f"chunk_id {chunk_id} no action needed")
+            continue
+
+        reinit = False
+        for n in range(rank):
```

Comment:
I think this could be done with better performance via `reinit = np.any(chunk_coord[:rank] < shape_update[:rank])`

Reply:
not sure how much better perorming it will be, but your method is more readable.  changed.